# TradingView Screener를 활용한 주식 종목 스크리닝

이 노트북에서는 `tradingview-screener` 라이브러리를 사용하여 다양한 조건으로 주식 종목을 스크리닝합니다.

## 주요 기능
- 기본 스크리닝 (종가, 거래량, 시가총액)
- 거래량 기반 스크리닝
- 기술적 지표 스크리닝 (MACD, RSI)
- 모멘텀 스크리닝
- 가치 투자 스크리닝


## 1. 라이브러리 설치 및 임포트


In [4]:
# 필요한 라이브러리 설치
# !pip install tradingview-screener pandas matplotlib seaborn


In [5]:
# 라이브러리 임포트
from tradingview_screener import Query, col
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# 한글 폰트 설정 (맥OS)
plt.rcParams['font.family'] = 'AppleGothic'
plt.rcParams['axes.unicode_minus'] = False

# 판다스 출력 설정
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.width', None)

# Seaborn 스타일 설정
sns.set_theme(style='whitegrid', palette='husl')

print("라이브러리 로딩 완료!")


라이브러리 로딩 완료!


## 2. 기본 스크리닝

가장 기본적인 스크리닝 - 종목명, 종가, 거래량, 시가총액 조회


In [6]:
# 기본 스크리닝: 상위 50개 종목
count, df_basic = (
    Query()
    .select('name', 'close', 'volume', 'market_cap_basic', 'change')
    .get_scanner_data()
)

print(f"전체 {count}개 종목 중 {len(df_basic)}개 조회됨")
df_basic.head(10)


전체 19256개 종목 중 50개 조회됨


ticker   name   close     volume  market_cap_basic    change
0      AMEX:SPY    SPY  681.53   62953844               NaN  0.185221
1    NASDAQ:QQQ    QQQ  622.00   54541216               NaN  0.782604
2   NASDAQ:NVDA   NVDA  181.46  182625215      4.409478e+12  0.855936
3   NASDAQ:TSLA   TSLA  429.24   69331263      1.427575e+12 -0.209234
4   NASDAQ:AAPL   AAPL  286.19   53669494      4.228844e+12  1.091487
5  NASDAQ:GOOGL  GOOGL  315.81   35854612      3.812127e+12  0.292166
6   NASDAQ:AMZN   AMZN  234.42   45784575      2.506000e+12  0.230888
7   NASDAQ:MSFT   MSFT  490.00   19562722      3.641865e+12  0.669762
8    NASDAQ:AMD    AMD  215.24   42001820      3.504196e+11 -2.056789
9   NASDAQ:AVGO   AVGO  381.57   22206602      1.801913e+12 -1.168152

## 3. 거래량 기반 스크리닝

상대 거래량(10일 평균 대비)이 높은 종목을 찾습니다.


In [7]:
# 거래량 스크리닝: 상대 거래량 1.5배 이상
count, df_volume = (
    Query()
    .select('name', 'close', 'volume', 'relative_volume_10d_calc', 'change', 'market_cap_basic')
    .where(
        col('relative_volume_10d_calc') > 1.5,
        col('market_cap_basic') > 100_000_000  # 시가총액 1억 달러 이상
    )
    .order_by('relative_volume_10d_calc', ascending=False)
    .limit(30)
    .get_scanner_data()
)

print(f"상대 거래량 1.5배 이상: {count}개 종목 중 {len(df_volume)}개 조회")
df_volume


상대 거래량 1.5배 이상: 1299개 종목 중 30개 조회


ticker   name     close    volume  relative_volume_10d_calc  \
0      OTC:DBOEF  DBOEF  263.8000    142877                332.194838   
1      OTC:KPELY  KPELY   15.8400    348396                227.605671   
2      OTC:LSRCF  LSRCF  170.6200    101211                162.509634   
3      OTC:STMEF  STMEF   23.5150    180008                127.619993   
4      OTC:HLPPY  HLPPY    5.8530    277689                105.229073   
5      OTC:GGDVY  GGDVY   49.2500     21694                105.157538   
6      OTC:FRFXF  FRFXF   18.0000     86023                103.692141   
7      OTC:ZHEXF  ZHEXF    1.0040    928718                 98.709479   
8      OTC:DISPF  DISPF  277.8500      2514                 91.418182   
9      OTC:DWLAF  DWLAF    1.0600     71750                 89.788512   
10     OTC:CHEAF  CHEAF    0.5800    678700                 86.673903   
11     OTC:LLDTF  LLDTF    1.2850   2034000                 76.595456   
12     OTC:VLPNY  VLPNY    8.7000    106111                 72.668812   
13       NYSE:EB     EB    4.4300  29467518                 71.746614   
14     OTC:SNPHF  SNPHF    9.5080     62682                 67.786309   
15      OTC:DBIN   DBIN   38.0000    140200                 67.058880   
16     OTC:ENGGF  ENGGF   16.1966     72090                 64.637317   
17     OTC:YAHOF  YAHOF    2.7100    428613                 63.006306   
18      NYSE:RZB    RZB   24.8700    826643                 62.177919   
19     OTC:FMXUF  FMXUF    9.4800     20000                 57.388809   
20     OTC:RTMVF  RTMVF    7.0000     88577                 49.261443   
21  NASDAQ:RAAQU  RAAQU   10.5000     30000                 48.653909   
22     OTC:CEBCF  CEBCF    0.4600   1080231                 46.729666   
23  NASDAQ:RDAGU  RDAGU   10.4500    105000                 42.717657   
24     OTC:PUTRY  PUTRY    4.8500     34263                 39.528150   
25      OTC:FNBT   FNBT   36.5000    236893                 39.491381   
26     OTC:LRCDF  LRCDF   28.6680     22943                 39.454858   
27     OTC:WTBDY  WTBDY    8.0600    436784                 34.812660   
28     OTC:PGPKY  PGPKY    4.8700     26144                 33.944430   
29     OTC:SWDCF  SWDCF    3.6500     13500                 32.081749   

       change  market_cap_basic  
0   -0.151400      4.796445e+10  
1   -0.251889      1.385248e+10  
2    0.134984      1.538616e+10  
3    1.335919      2.098310e+10  
4    1.614583      5.854166e+09  
5   -1.024920      6.343817e+09  
6    1.838755      4.081744e+10  
7    1.414141      6.062267e+09  
8   -4.600858      3.012762e+10  
9    2.912621      1.389804e+09  
10  14.579218      1.472062e+10  
11  11.739130      7.556114e+10  
12   2.352941      7.820873e+09  
13  78.629032      4.327009e+08  
14  -5.907966      3.060102e+09  
15  -0.783290      4.243840e+08  
16   1.070827      4.235219e+09  
17   6.692913      1.828305e+10  
18   0.100624      1.266240e+10  
19  -0.940439      1.985608e+10  
20  -1.408451      5.341751e+09  
21   0.000000      2.355200e+08  
22  -4.166667      2.882919e+10  
23  -1.877934      3.926250e+08  
24  -6.730769      2.787570e+10  
25  -2.275770      4.482953e+08  
26  19.003736      1.277256e+09  
27  -1.104294      5.527751e+09  
28 -10.147601      5.796121e+09  
29  -5.194805      5.834195e+08

## 4. 기술적 지표 스크리닝 (MACD, RSI)

MACD 골든 크로스와 RSI 조건을 활용한 스크리닝


In [8]:
# MACD 골든 크로스 스크리닝
count, df_macd = (
    Query()
    .select('name', 'close', 'change', 'volume', 'MACD.macd', 'MACD.signal', 'RSI')
    .where(
        col('market_cap_basic') > 50_000_000,
        col('MACD.macd') >= col('MACD.signal'),  # MACD가 시그널 위에 있음 (골든 크로스)
        col('RSI').between(30, 70)  # 과매수/과매도 아닌 구간
    )
    .order_by('volume', ascending=False)
    .limit(30)
    .get_scanner_data()
)

print(f"MACD 골든 크로스 종목: {count}개 중 {len(df_macd)}개 조회")
df_macd


MACD 골든 크로스 종목: 5339개 중 30개 조회


ticker   name     close     change     volume  MACD.macd  \
0    NASDAQ:BYND   BYND    1.2900  -3.731343  233167768  -0.166014   
1    NASDAQ:INTC   INTC   43.4700   8.647838  167195636   0.828437   
2    NASDAQ:BITF   BITF    3.1000  -5.487805   93787257  -0.182109   
3    NASDAQ:ONDS   ONDS    8.0700   6.464380   90635527   0.259068   
4    NASDAQ:TSLA   TSLA  429.2400  -0.209234   69331263  -2.581048   
5     NASDAQ:AAL    AAL   14.2400   2.005731   64579405   0.331858   
6    NASDAQ:ASST   ASST    1.0350  -1.428571   64148284  -0.168194   
7      AMEX:BMNR   BMNR   31.9100  10.262612   60475192  -4.416138   
8      NYSE:BBAI   BBAI    5.7900  -4.297521   59959702  -0.177434   
9    NASDAQ:SOFI   SOFI   29.5100   1.653462   59744761   0.069753   
10   NASDAQ:MARA   MARA   11.9100   3.385417   46125092  -1.669052   
11      NYSE:BAC    BAC   53.1900  -0.093914   46073513   0.248236   
12  NASDAQ:CMCSA  CMCSA   27.0200   1.655380   45781414  -0.618275   
13     NYSE:ACHR   ACHR    7.6500   2.546917   45457186  -0.741943   
14      AMEX:DNN    DNN    2.5900   2.777778   45261293  -0.068073   
15      NYSE:PFE    PFE   25.1500  -0.474871   43826780   0.150891   
16    NASDAQ:CAN    CAN    0.9672   2.046845   40891514  -0.094011   
17   NASDAQ:CLSK   CLSK   13.7100  -2.627841   40094016  -0.799467   
18     NYSE:ABEV   ABEV    2.5800   0.389105   38633711   0.064223   
19   NASDAQ:BTBT   BTBT    2.2400  -0.444444   38280287  -0.304816   
20   NASDAQ:MRVL   MRVL   92.8900   1.964874   38225532   0.625045   
21     NYSE:ITUB   ITUB    7.8900   2.600780   36312821   0.117049   
22   NASDAQ:CIFR   CIFR   17.6400 -10.411376   36127479   0.130539   
23  NASDAQ:GOOGL  GOOGL  315.8100   0.292166   35854612  14.168940   
24      AMEX:BTG    BTG    4.4300  -3.695652   35125057  -0.050574   
25   NASDAQ:RIVN   RIVN   17.2300   0.407925   34438794   0.662234   
26       NYSE:NU     NU   17.6100   0.743707   34356463   0.425435   
27   NASDAQ:GRAB   GRAB    5.2600  -0.189753   33420441  -0.149205   
28   NASDAQ:APLD   APLD   28.0500  -0.567175   32943970  -0.918873   
29      NYSE:EXK    EXK    9.2400  -1.702128   32919947   0.311885   

    MACD.signal        RSI  
0     -0.230615  50.086889  
1      0.189735  69.769342  
2     -0.242096  47.206277  
3      0.075703  55.172941  
4     -4.602919  52.095014  
5      0.199193  64.219356  
6     -0.208491  39.164425  
7     -4.848890  43.095594  
8     -0.214915  45.278495  
9     -0.105990  54.496481  
10    -1.794201  37.654203  
11     0.196902  55.630307  
12    -0.702092  43.404938  
13    -0.814933  39.872866  
14    -0.090309  50.588625  
15     0.130921  50.494009  
16    -0.098492  44.572727  
17    -1.201692  49.383807  
18     0.064121  68.831362  
19    -0.311655  35.834263  
20    -0.178228  60.069049  
21     0.098528  66.361070  
22    -0.099512  50.235978  
23    12.626971  68.705978  
24    -0.144412  51.902917  
25     0.510681  65.618815  
26     0.234183  69.695431  
27    -0.152202  42.699431  
28    -1.350366  53.783832  
29     0.092513  61.619948

## 5. 모멘텀 스크리닝

상승 중이면서 거래량이 증가하는 종목 찾기


In [9]:
# 모멘텀 스크리닝
count, df_momentum = (
    Query()
    .select('name', 'close', 'change', 'volume', 'relative_volume_10d_calc', 
            'market_cap_basic', 'Perf.W', 'Perf.1M')  # 주간, 월간 수익률
    .where(
        col('change') > 3,  # 오늘 3% 이상 상승
        col('relative_volume_10d_calc') > 1.2,  # 거래량 1.2배 이상
        col('market_cap_basic') > 100_000_000  # 시가총액 1억 달러 이상
    )
    .order_by('change', ascending=False)
    .limit(25)
    .get_scanner_data()
)

print(f"모멘텀 상승 종목: {count}개 중 {len(df_momentum)}개 조회")
df_momentum


모멘텀 상승 종목: 289개 중 25개 조회


ticker   name     close        change    volume  \
0     OTC:AMVIF  AMVIF   42.5000  4.249990e+07       647   
1     OTC:ACEHF  ACEHF    0.0323  2.230000e+02     10000   
2     OTC:NFGRF  NFGRF    0.2127  9.934396e+01      5000   
3     OTC:AGTEF  AGTEF    0.0202  9.423077e+01     20000   
4       NYSE:EB     EB    4.4300  7.862903e+01  29467518   
5     OTC:BNPJY  BNPJY    2.5000  5.625000e+01       200   
6     OTC:MKNGF  MKNGF    0.6000  5.584416e+01     19173   
7     OTC:TRGNF  TRGNF    1.0800  5.428571e+01      2000   
8     OTC:COOSF  COOSF   13.9650  5.229008e+01      5878   
9     OTC:MRTPY  MRTPY    7.4500  5.050505e+01      1167   
10   NASDAQ:THH    THH   22.4200  4.966622e+01    238679   
11    OTC:HWLDF  HWLDF    4.4700  4.900000e+01      1100   
12    OTC:ZHAOF  ZHAOF    4.0500  4.303373e+01      2450   
13    OTC:BXPHF  BXPHF    0.0783  3.617391e+01     69400   
14    OTC:NIPMY  NIPMY   21.5000  2.647059e+01       300   
15    OTC:BAWAY  BAWAY   34.1150  2.240761e+01      2300   
16   NASDAQ:MDB    MDB  401.9900  2.223371e+01   9487425   
17    OTC:WMWWF  WMWWF    0.0605  2.104842e+01    143440   
18    OTC:LSHGF  LSHGF    1.6900  2.071429e+01     45932   
19  NASDAQ:FBYD   FBYD   20.8000  2.044007e+01    732481   
20    OTC:QUCOF  QUCOF   10.1800  1.976471e+01     23000   
21    OTC:LRCDF  LRCDF   28.6680  1.900374e+01     22943   
22  NASDAQ:GUTS   GUTS    1.8400  1.794872e+01  11879113   
23   NASDAQ:TMC    TMC    7.4000  1.774065e+01  23049597   
24    OTC:LINIF  LINIF    0.2458  1.738300e+01     27800   

    relative_volume_10d_calc  market_cap_basic        Perf.W     Perf.1M  
0                   1.833381      4.252167e+09  4.249990e+07    3.658537  
1                   2.433327      5.520322e+08  2.230000e+02  822.857143  
2                   6.572893      1.832429e+08  9.934396e+01   99.343955  
3                   1.514211      2.322828e+08  4.975124e-01   -3.579952  
4                  71.746614      4.327009e+08  8.155738e+01   94.298246  
5                   1.312336      1.476843e+09  5.625000e+01   56.250000  
6                   1.210455      2.083157e+08  8.108108e+00  -25.000000  
7                   4.257131      1.430077e+08  5.428571e+01   54.285714  
8                   4.826340      2.360411e+08  3.731563e+01   72.194821  
9                   2.147589      4.943515e+08  5.050505e+01   50.505051  
10                  6.499601      1.122037e+09  1.463736e+02  142.640693  
11                  1.368329      1.337577e+10  4.900000e+01   49.000000  
12                  1.487373      1.434669e+10  4.303373e+01   43.033728  
13                  1.824400      1.543027e+08  3.617391e+01  -13.000000  
14                  1.850709      4.353337e+09           NaN   26.470588  
15                  2.348137      1.037738e+10  2.240761e+01   22.407607  
16                  5.162453      3.270465e+10  2.197409e+01   13.396333  
17                  2.755346      2.320023e+08  9.403255e+00   51.250000  
18                  1.282096      1.676474e+09  2.518519e+01    5.625000  
19                 10.638362      2.518090e+09  5.745647e+01   94.029851  
20                  1.536961      4.072000e+09  1.634286e+01   16.342857  
21                 39.454858      1.277256e+09  2.199149e+01   23.409384  
22                  4.840057      2.521617e+08  2.517007e+01   49.593496  
23                  2.272952      3.059841e+09  3.909774e+01    3.351955  
24                  2.206629      4.038374e+08  1.738300e+01   -2.730510

## 6. 시가총액별 스크리닝


In [10]:
# 대형주 스크리닝 (시가총액 100억 달러 이상)
count, df_large_cap = (
    Query()
    .select('name', 'close', 'change', 'volume', 'market_cap_basic', 'sector')
    .where(
        col('market_cap_basic') > 10_000_000_000  # 100억 달러 이상
    )
    .order_by('market_cap_basic', ascending=False)
    .limit(30)
    .get_scanner_data()
)

print(f"대형주 종목: {count}개 중 {len(df_large_cap)}개 조회")
df_large_cap


대형주 종목: 2245개 중 30개 조회


ticker    name        close    change     volume  market_cap_basic  \
0    NASDAQ:NVDA    NVDA     181.4600  0.855936  182625215      4.409478e+12   
1    NASDAQ:AAPL    AAPL     286.1900  1.091487   53669494      4.228844e+12   
2   NASDAQ:GOOGL   GOOGL     315.8100  0.292166   35854612      3.812127e+12   
3    NASDAQ:GOOG    GOOG     316.0200  0.285605   24668179      3.811877e+12   
4    NASDAQ:MSFT    MSFT     490.0000  0.669762   19562722      3.641865e+12   
5    NASDAQ:AMZN    AMZN     234.4200  0.230888   45784575      2.506000e+12   
6    NASDAQ:AVGO    AVGO     381.5700 -1.168152   22206602      1.801913e+12   
7    NASDAQ:META    META     647.1000  0.972116   11640836      1.631033e+12   
8    NASDAQ:TSLA    TSLA     429.2400 -0.209234   69331263      1.427575e+12   
9       NYSE:TSM     TSM     292.0900  1.532953   10745439      1.162580e+12   
10    NYSE:BRK.A   BRK.A  760100.0500 -0.139332        270      1.094715e+12   
11    NYSE:BRK.B   BRK.B     506.6500 -0.373611    3378009      1.092147e+12   
12      NYSE:LLY     LLY    1046.1200 -1.112592    3279957      9.889848e+11   
13      NYSE:WMT     WMT     112.4100  0.789025   18711768      8.962282e+11   
14   NYSE:JPM/PL  JPM/PL      19.5900 -0.709579     281716      8.409613e+11   
15   NYSE:JPM/PC  JPM/PC      25.2374 -0.168513      87105      8.409613e+11   
16   NYSE:JPM/PK  JPM/PK      19.2800 -0.464636     325476      8.409613e+11   
17   NYSE:JPM/PJ  JPM/PJ      20.1500 -0.493827      93885      8.409613e+11   
18   NYSE:JPM/PM  JPM/PM      17.9800 -1.208791     930194      8.409613e+11   
19   NYSE:JPM/PD  JPM/PD      24.9700 -0.199840     117989      8.409613e+11   
20      NYSE:JPM     JPM     307.8800 -0.336657    7249043      8.381301e+11   
21     OTC:TCEHY   TCEHY      79.1600 -0.814434    2040968      7.205560e+11   
22     OTC:TCTZF   TCTZF      77.7200 -2.435350      14361      7.043092e+11   
23        NYSE:V       V     329.6200 -0.233058    8201830      6.309241e+11   
24     NYSE:ORCL    ORCL     201.1000  0.079626   19664465      5.732944e+11   
25      NYSE:JNJ     JNJ     205.4200  0.038960    8393565      4.949174e+11   
26       NYSE:MA      MA     545.9100  0.356637    3076100      4.902249e+11   
27      NYSE:XOM     XOM     115.3800 -1.071765   14607694      4.865765e+11   
28   NASDAQ:NFLX    NFLX     109.3500  0.201594   25761972      4.633513e+11   
29     OTC:ASMLF   ASMLF    1103.0480  2.661068       6531      4.213476e+11   

                   sector  
0   Electronic Technology  
1   Electronic Technology  
2     Technology Services  
3     Technology Services  
4     Technology Services  
5            Retail Trade  
6   Electronic Technology  
7     Technology Services  
8       Consumer Durables  
9   Electronic Technology  
10                Finance  
11                Finance  
12      Health Technology  
13           Retail Trade  
14                Finance  
15                Finance  
16                Finance  
17                Finance  
18                Finance  
19                Finance  
20                Finance  
21    Technology Services  
22    Technology Services  
23                Finance  
24    Technology Services  
25      Health Technology  
26                Finance  
27        Energy Minerals  
28    Technology Services  
29  Electronic Technology

In [11]:
# 중형주 스크리닝 (시가총액 10억 ~ 100억 달러)
count, df_mid_cap = (
    Query()
    .select('name', 'close', 'change', 'volume', 'market_cap_basic', 'sector')
    .where(
        col('market_cap_basic').between(1_000_000_000, 10_000_000_000)
    )
    .order_by('change', ascending=False)
    .limit(30)
    .get_scanner_data()
)

print(f"중형주 종목: {count}개 중 {len(df_mid_cap)}개 조회")
df_mid_cap


중형주 종목: 3130개 중 30개 조회


ticker   name      close        change  volume  market_cap_basic  \
0     OTC:AMVIF  AMVIF   42.50000  4.249990e+07     647      4.252167e+09   
1     OTC:YRCCF  YRCCF   17.19000  1.951411e+02     400      1.177163e+09   
2     OTC:SHPHF  SHPHF    0.20770  1.077000e+02     106      1.937841e+09   
3     OTC:LGBOF  LGBOF    1.95000  9.500000e+01     100      2.509603e+09   
4     OTC:MGHTF  MGHTF    3.90000  9.259259e+01    1500      5.527833e+09   
5     OTC:UNEGF  UNEGF    0.07000  7.283951e+01   10000      1.818850e+09   
6      OTC:ALTB   ALTB    8.00000  6.000000e+01     100      7.968800e+09   
7     OTC:BNPJY  BNPJY    2.50000  5.625000e+01     200      1.476843e+09   
8     OTC:PUPOF  PUPOF   18.39000  5.427464e+01    1690      6.790875e+09   
9     OTC:BUMTF  BUMTF    1.15000  5.333333e+01    2812      1.994266e+09   
10    OTC:NISUF  NISUF    8.81000  5.189655e+01     300      2.742017e+09   
11    OTC:FJLLF  FJLLF   22.25000  5.084746e+01     150      1.912659e+09   
12   NASDAQ:THH    THH   22.42000  4.966622e+01  238679      1.122037e+09   
13    OTC:AYAAY  AYAAY    9.14000  3.246377e+01     100      5.126189e+09   
14    OTC:CCOHF  CCOHF    1.16000  3.107345e+01     112      6.127371e+09   
15    OTC:DFILF  DFILF    3.22340  3.066072e+01    1300      4.363358e+09   
16    OTC:TPGVF  TPGVF    0.13975  2.915896e+01     150      1.120510e+09   
17    OTC:PTNRF  PTNRF   12.00000  2.725345e+01     104      2.245075e+09   
18    OTC:KSPHF  KSPHF   31.90000  2.709163e+01     400      1.303010e+09   
19    OTC:RGLXY  RGLXY    4.10000  2.673879e+01     197      6.071444e+09   
20    OTC:NIPMY  NIPMY   21.50000  2.647059e+01     300      4.353337e+09   
21    OTC:VRBCF  VRBCF  436.32000  2.610405e+01       2      3.654555e+09   
22    OTC:DLPTF  DLPTF    0.10000  2.500000e+01    1500      4.404271e+09   
23    OTC:WILWY  WILWY   18.15000  2.330163e+01     250      3.862377e+09   
24    OTC:YASKF  YASKF   24.00000  2.176560e+01     300      6.273924e+09   
25    OTC:SAEYY  SAEYY    8.85000  2.123288e+01     477      1.545541e+09   
26    OTC:LSHGF  LSHGF    1.69000  2.071429e+01   45932      1.676474e+09   
27    OTC:TCMFF  TCMFF    2.35000  2.051282e+01     500      5.060916e+09   
28  NASDAQ:FBYD   FBYD   20.80000  2.044007e+01  732481      2.518090e+09   
29    OTC:PAFRY  PAFRY   28.95000  1.987578e+01     143      3.194485e+09   

                    sector  
0   Producer Manufacturing  
1      Industrial Services  
2        Health Technology  
3      Non-Energy Minerals  
4                Utilities  
5          Energy Minerals  
6             Retail Trade  
7          Energy Minerals  
8                Utilities  
9       Process Industries  
10   Consumer Non-Durables  
11      Process Industries  
12       Consumer Services  
13                 Finance  
14     Industrial Services  
15            Retail Trade  
16       Health Technology  
17          Communications  
18       Health Technology  
19       Consumer Services  
20   Consumer Non-Durables  
21       Health Technology  
22          Transportation  
23          Transportation  
24  Producer Manufacturing  
25   Distribution Services  
26           Miscellaneous  
27          Communications  
28     Industrial Services  
29     Non-Energy Minerals

## 7. 가치 투자 스크리닝

낮은 P/E 비율과 배당 수익률 기준 스크리닝


In [12]:
# 가치 투자 스크리닝
count, df_value = (
    Query()
    .select('name', 'close', 'price_earnings_ttm', 'dividend_yield_recent', 
            'price_book_ratio', 'market_cap_basic', 'sector')
    .where(
        col('price_earnings_ttm').between(5, 15),  # P/E 5~15
        col('dividend_yield_recent') > 2,  # 배당수익률 2% 이상
        col('market_cap_basic') > 1_000_000_000  # 시가총액 10억 달러 이상
    )
    .order_by('dividend_yield_recent', ascending=False)
    .limit(25)
    .get_scanner_data()
)

print(f"가치 투자 종목: {count}개 중 {len(df_value)}개 조회")
df_value


가치 투자 종목: 1024개 중 25개 조회


ticker     name    close  price_earnings_ttm  dividend_yield_recent  \
0      OTC:AKRYY    AKRYY   5.7000            5.795628              54.533108   
1      OTC:MBRFY    MBRFY   3.4600            6.229744              26.149770   
2       NYSE:ORC      ORC   7.2100           14.875181              19.972261   
3      OTC:UNEGF    UNEGF   0.0700           10.949527              18.181818   
4       NYSE:GOF      GOF  12.2000           12.808399              17.909836   
5       AMEX:CLM      CLM   8.2800            8.625000              17.608696   
6        NYSE:EC       EC   9.9100            7.508713              16.891040   
7       NYSE:MNR      MNR  12.0800           12.584644              16.059603   
8      NYSE:IIPR     IIPR  49.8700           11.789319              15.239623   
9   NYSE:IIPR/PA  IIPR/PA  24.4400            5.777641              15.239623   
10       NYSE:DX       DX  13.8300            9.343332              14.750542   
11      NYSE:PDI      PDI  18.0900            7.697872              14.626865   
12   NASDAQ:TRIN     TRIN  14.5600            6.662091              14.010988   
13  NASDAQ:TRINZ    TRINZ  25.4964           11.666163              14.010988   
14  NASDAQ:TRINI    TRINI  25.5600           11.695264              14.010988   
15     NYSE:BWLP     BWLP  11.9900           10.011566              13.604707   
16      NYSE:JBS      JBS  14.7400            9.692905              13.600535   
17     NASDAQ:XP       XP  20.0200           12.115710              13.136664   
18     NYSE:GSBD     GSBD   9.8900            8.673156              12.942365   
19     OTC:GQPIL    GQPIL   0.8000            5.033438              12.910446   
20      NYSE:JFR      JFR   7.9300           11.046107              12.862547   
21     OTC:LCSHF    LCSHF   7.6300            7.786472              12.850957   
22      NYSE:UAN      UAN  94.8100            7.877071              12.572514   
23   NASDAQ:MFIC     MFIC  12.1000           11.322167              12.561983   
24      NYSE:PMT      PMT  12.8100           13.978612              12.490242   

    price_book_ratio  market_cap_basic                 sector  
0           1.376659      1.467902e+09              Utilities  
1           6.677474      4.889898e+09  Consumer Non-Durables  
2           0.891111      1.089293e+09                Finance  
3           1.050753      1.818850e+09        Energy Minerals  
4           1.067537      1.987849e+09          Miscellaneous  
5           1.177833      2.092828e+09          Miscellaneous  
6           1.072388      2.041403e+10        Energy Minerals  
7           1.192182      2.034547e+09        Energy Minerals  
8           0.738804      1.398066e+09                Finance  
9           0.362069      1.398066e+09                Finance  
10          1.084885      2.037886e+09                Finance  
11          1.068748      6.817181e+09          Miscellaneous  
12          1.091034      1.101950e+09          Miscellaneous  
13          1.910538      1.103463e+09          Miscellaneous  
14          1.915304      1.103463e+09          Miscellaneous  
15          1.009226      1.810904e+09         Transportation  
16          3.786779      1.523611e+10  Consumer Non-Durables  
17          3.316217      1.054938e+10                Finance  
18          0.737629      1.128579e+09          Miscellaneous  
19          5.695206      3.547172e+09                Finance  
20          0.898605      1.059042e+09          Miscellaneous  
21          1.230543      1.861796e+09                Finance  
22          3.419356      1.002107e+09     Process Industries  
23          0.807849      1.130840e+09                Finance  
24          0.796474      1.122514e+09                Finance

## 8. 사용자 정의 스크리닝 함수


In [13]:
def custom_screener(
    min_market_cap: int = 100_000_000,
    max_market_cap: int = None,
    min_volume: int = 100_000,
    min_change: float = None,
    max_change: float = None,
    min_rel_volume: float = None,
    min_rsi: float = None,
    max_rsi: float = None,
    macd_bullish: bool = False,
    sort_by: str = 'volume',
    ascending: bool = False,
    limit: int = 50
):
    """
    사용자 정의 스크리닝 함수
    
    Parameters:
    -----------
    min_market_cap : int - 최소 시가총액 (달러)
    max_market_cap : int - 최대 시가총액 (달러)
    min_volume : int - 최소 거래량
    min_change : float - 최소 변동률 (%)
    max_change : float - 최대 변동률 (%)
    min_rel_volume : float - 최소 상대 거래량 배수
    min_rsi : float - 최소 RSI
    max_rsi : float - 최대 RSI
    macd_bullish : bool - MACD 골든크로스 필터 적용 여부
    sort_by : str - 정렬 기준 컬럼
    ascending : bool - 오름차순 정렬 여부
    limit : int - 조회할 종목 수
    
    Returns:
    --------
    pandas.DataFrame - 스크리닝 결과
    """
    
    # 조건 목록 생성
    conditions = []
    
    # 시가총액 조건
    if max_market_cap:
        conditions.append(col('market_cap_basic').between(min_market_cap, max_market_cap))
    else:
        conditions.append(col('market_cap_basic') > min_market_cap)
    
    # 거래량 조건
    conditions.append(col('volume') > min_volume)
    
    # 변동률 조건
    if min_change is not None and max_change is not None:
        conditions.append(col('change').between(min_change, max_change))
    elif min_change is not None:
        conditions.append(col('change') > min_change)
    elif max_change is not None:
        conditions.append(col('change') < max_change)
    
    # 상대 거래량 조건
    if min_rel_volume is not None:
        conditions.append(col('relative_volume_10d_calc') > min_rel_volume)
    
    # RSI 조건
    if min_rsi is not None and max_rsi is not None:
        conditions.append(col('RSI').between(min_rsi, max_rsi))
    elif min_rsi is not None:
        conditions.append(col('RSI') > min_rsi)
    elif max_rsi is not None:
        conditions.append(col('RSI') < max_rsi)
    
    # MACD 조건
    if macd_bullish:
        conditions.append(col('MACD.macd') >= col('MACD.signal'))
    
    # 쿼리 실행
    query = (
        Query()
        .select('name', 'close', 'change', 'volume', 'relative_volume_10d_calc',
                'market_cap_basic', 'RSI', 'MACD.macd', 'MACD.signal', 'sector')
        .where(*conditions)
        .order_by(sort_by, ascending=ascending)
        .limit(limit)
    )
    
    count, df = query.get_scanner_data()
    
    print(f"조건에 맞는 종목: {count}개 중 {len(df)}개 조회")
    
    return df


In [14]:
# 사용자 정의 스크리닝 예시
# 조건: 중형주, 2% 이상 상승, 거래량 증가, MACD 골든크로스

df_custom = custom_screener(
    min_market_cap=500_000_000,
    max_market_cap=10_000_000_000,
    min_volume=500_000,
    min_change=2.0,
    min_rel_volume=1.2,
    macd_bullish=True,
    sort_by='change',
    ascending=False,
    limit=20
)

df_custom


조건에 맞는 종목: 64개 중 20개 조회


ticker  name   close     change    volume  relative_volume_10d_calc  \
0   NASDAQ:FBYD  FBYD   20.80  20.440069    732481                 10.638362   
1    NASDAQ:TMC   TMC    7.40  17.740652  23049597                  2.272952   
2   NASDAQ:VERA  VERA   37.25  12.878788   4354087                  3.183776   
3     NYSE:EVEX  EVEX    4.67  11.990408   3083723                  2.642696   
4     NYSE:CTRI  CTRI   24.46  10.229833   4281515                  2.248378   
5     NYSE:CSAN  CSAN    5.34   8.757637   2654931                  1.827894   
6     NYSE:ADCT  ADCT    4.60   8.490566   1309400                  1.917060   
7   NASDAQ:SAIA  SAIA  314.50   7.282961    973906                  2.293188   
8   NASDAQ:OMCL  OMCL   38.46   6.478405    622874                  1.668729   
9   NASDAQ:TRON  TRON    2.00   6.382979   1453338                  1.293673   
10      NYSE:TE    TE    4.74   6.040268  13323488                  1.207901   
11    NYSE:INSP  INSP  133.93   6.007599   2752251                  1.507284   
12  NASDAQ:ATAT  ATAT   42.69   5.589908   3256496                  2.390339   
13  NASDAQ:FWDI  FWDI    8.29   5.470738    901513                  1.503506   
14  NASDAQ:WERN  WERN   27.18   5.308020   1583083                  2.047607   
15  NASDAQ:CAMT  CAMT  110.90   5.218216   1168855                  2.578640   
16    NYSE:ASAN  ASAN   13.39   4.855129   9292854                  2.143993   
17    NYSE:ONTO  ONTO  154.01   4.768707    898555                  1.299895   
18     NYSE:SPR   SPR   37.54   4.714086   2931418                  2.374931   
19  NASDAQ:PLAB  PLAB   23.49   4.679144   2194890                  3.600971   

    market_cap_basic        RSI  MACD.macd  MACD.signal  \
0       2.518090e+09  69.851372   1.724212     0.962943   
1       3.059841e+09  61.835256  -0.085690    -0.346700   
2       2.381320e+09  71.757235   1.507663     0.663915   
3       1.626582e+09  64.071311  -0.052899    -0.117483   
4       2.435712e+09  70.021610   0.476190     0.126424   
5       4.535634e+09  65.777467   0.098139     0.065678   
6       5.698347e+08  64.090673   0.074329     0.034000   
7       8.379110e+09  66.779360   0.312366    -4.967937   
8       1.725951e+09  69.578930   1.175447     1.089960   
9       5.142308e+08  43.973209  -0.219364    -0.262171   
10      1.006614e+09  65.636404   0.158399     0.024883   
11      3.891528e+09  79.835743  12.552086     7.848889   
12      5.586858e+09  65.123672   0.325676     0.040759   
13      7.167490e+08  32.464338  -2.272681    -2.699686   
14      1.626188e+09  61.662123  -0.178817    -0.538043   
15      5.075478e+09  55.310935  -2.136578    -3.162567   
16      3.163347e+09  56.061556  -0.258304    -0.387487   
17      7.548209e+09  67.634082   3.574830     1.299108   
18      4.411821e+09  57.316569  -0.350667    -0.607692   
19      1.386019e+09  57.851077  -0.101426    -0.370565   

                    sector  
0      Industrial Services  
1      Non-Energy Minerals  
2        Health Technology  
3    Electronic Technology  
4                Utilities  
5                Utilities  
6        Health Technology  
7           Transportation  
8      Technology Services  
9        Consumer Durables  
10  Producer Manufacturing  
11       Health Technology  
12       Consumer Services  
13       Consumer Durables  
14          Transportation  
15   Electronic Technology  
16     Technology Services  
17   Electronic Technology  
18   Electronic Technology  
19   Electronic Technology

## 9. 결과 저장


In [15]:
# 결과를 CSV로 저장
from datetime import datetime

timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')

# 예시: 모멘텀 스크리닝 결과 저장
if 'df_momentum' in dir() and not df_momentum.empty:
    filename = f'momentum_screening_{timestamp}.csv'
    df_momentum.to_csv(filename, index=False, encoding='utf-8-sig')
    print(f"저장 완료: {filename}")


저장 완료: momentum_screening_20251203_174517.csv


## 사용 가능한 필드 목록 (예시)

| 필드명 | 설명 |
|--------|------|
| name | 종목명 |
| close | 종가 |
| open | 시가 |
| high | 고가 |
| low | 저가 |
| volume | 거래량 |
| change | 변동률 (%) |
| market_cap_basic | 시가총액 |
| relative_volume_10d_calc | 10일 평균 대비 상대 거래량 |
| RSI | RSI (14일) |
| MACD.macd | MACD 라인 |
| MACD.signal | MACD 시그널 |
| price_earnings_ttm | P/E 비율 (TTM) |
| dividend_yield_recent | 배당 수익률 |
| price_book_ratio | P/B 비율 |
| sector | 섹터 |
| industry | 산업 |
| Perf.W | 주간 수익률 |
| Perf.1M | 월간 수익률 |
| price_52_week_high | 52주 최고가 |
| price_52_week_low | 52주 최저가 |


---

# Part 2: Global Screener - 4가지 투자 전략

`global_screener_spec.md`에 정의된 4가지 투자 전략 스크리너입니다:
1. **Cyclical (경기민감형)** - 저 PBR, 저 EV/EBITDA
2. **Growth (고성장형)** - 높은 매출 성장률, 저 PEG
3. **Finance (금융/자산주)** - 극저 PBR, 높은 ROE, 배당
4. **Defensive (경기방어주)** - 안정적 영업이익률, FCF, 배당


In [135]:
# 공통 설정

# 기술 등급 기준 (Technical Rating)
# Recommend.All: -1(Strong Sell) ~ 1(Strong Buy)
# -1 ~ -0.5: Strong Sell, -0.5 ~ -0.1: Sell, -0.1 ~ 0.1: Neutral
# 0.1 ~ 0.5: Buy, 0.5 ~ 1: Strong Buy
TECH_RATING_BUY = 0.1
TECH_RATING_STRONG_BUY = 0.5

# 애널리스트 평점 기준 (Analyst Rating)
# 가중 평균 점수 (-2 ~ 2 스케일)
# Strong Buy=2, Buy=1, Hold=0, Sell=-1, Strong Sell=-2
ANALYST_SCORE_BUY = 0.5        # Buy 이상 기준
ANALYST_SCORE_STRONG_BUY = 1.0  # Strong Buy 기준

# 최소 애널리스트 수 (신뢰도 기준)
# 3명 이상: 유연한 기준 (소형주 포함)
MIN_ANALYST_COUNT = 3

# 섹터 매핑 (TradingView 영문 섹터명)
SECTORS = {
    'cyclical': [
        'Process Industries', 'Non-Energy Minerals', 'Producer Manufacturing',
        'Consumer Durables', 'Energy Minerals', 'Electronic Technology',
    ],
    'growth': [
        'Technology Services', 'Health Services', 'Commercial Services', 'Health Technology',
    ],
    'finance': ['Finance'],
    'defensive': ['Consumer Non-Durables', 'Utilities', 'Communications'],
}


def calculate_analyst_score(df):
    """
    애널리스트 평점 계산 함수
    점수 = (2×strong_buy + 1×buy + 0×hold - 1×sell - 2×strong_sell) / total
    """
    if df.empty:
        return df
    
    required_cols = ['recommendation_buy', 'recommendation_over', 'recommendation_hold',
                     'recommendation_under', 'recommendation_sell', 'recommendation_total']
    
    if not all(c in df.columns for c in required_cols):
        return df
    
    for c in required_cols:
        df[c] = df[c].fillna(0)
    
    df['analyst_score'] = (
        2 * df['recommendation_buy'] +
        1 * df['recommendation_over'] +
        0 * df['recommendation_hold'] +
        -1 * df['recommendation_under'] +
        -2 * df['recommendation_sell']
    ) / df['recommendation_total'].replace(0, 1)
    
    def get_rating(score, total):
        if pd.isna(score) or total == 0:
            return 'N/A'
        elif score >= 1.0:
            return 'Strong Buy'
        elif score >= 0.5:
            return 'Buy'
        elif score >= -0.5:
            return 'Hold'
        elif score >= -1.0:
            return 'Sell'
        else:
            return 'Strong Sell'
    
    df['analyst_rating'] = df.apply(lambda r: get_rating(r['analyst_score'], r['recommendation_total']), axis=1)
    return df


def filter_by_analyst(df, min_score=ANALYST_SCORE_BUY):
    """애널리스트 평점으로 필터링"""
    if df.empty or 'analyst_score' not in df.columns:
        return df
    return df[df['analyst_score'] >= min_score]


print("공통 설정 완료!")


공통 설정 완료!


## 전략 1: Cyclical (경기민감형)

**목표:** 자산 가치 대비 저평가되고, 현금 창출력이 좋은 기업

| 필터 | 조건 | 설명 |
|------|------|------|
| PBR | < 1 | 자산가치 대비 저평가 |
| EV/EBITDA | < 6 | 현금 창출력 대비 저평가 |
| 유동비율 | >= 1.5 | 경기 침체 시 버틸 현금 체력 |
| 애널리스트/기술 등급 | Buy 이상 | 전문가 평가 매수 이상 |


In [149]:
# Cyclical (경기민감형) 스크리너
count_cyclical, df_cyclical = (
    Query()
    .select(
        'name', 'close', 'change', 'volume', 'market_cap_basic',
        'sector', 'industry',
        'price_book_fq',               # PBR
        'enterprise_value_ebitda_ttm', # EV/EBITDA
        'current_ratio_fq',            # 유동비율
        'recommendation_mark',         # 애널리스트 평점 (문자열)
        'recommendation_hold',
        'recommendation_sell',
        'recommendation_buy',
        'recommendation_total',
        'recommendation_over',
        'recommendation_under',
        'Recommend.All',               # 기술 등급 (숫자)
    )
    .where(
        col('is_primary') == True,
        col('price_book_fq') < 1,
        col('price_book_fq') > 0,
        col('enterprise_value_ebitda_ttm') < 6,
        col('enterprise_value_ebitda_ttm') > 0,
        col('current_ratio_fq') >= 1.5,
        col('recommendation_total') >= MIN_ANALYST_COUNT,  # 최소 애널리스트 수
        col('Recommend.All') >= TECH_RATING_BUY,
    )
    .order_by('enterprise_value_ebitda_ttm', ascending=True)
    .limit(1000)
    .get_scanner_data()
)

# 애널리스트 점수 계산
df_cyclical = calculate_analyst_score(df_cyclical)

# 애널리스트 평점 Buy 이상 필터링
df_cyclical = filter_by_analyst(df_cyclical, ANALYST_SCORE_BUY)

# 섹터 필터링
if not df_cyclical.empty and 'sector' in df_cyclical.columns:
    df_cyclical_filtered = df_cyclical[df_cyclical['sector'].isin(SECTORS['cyclical'])]

print(f"📊 Cyclical 종목: {count_cyclical}개 중 {len(df_cyclical_filtered)}개 필터링됨")
print(f"조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)")
df_cyclical_filtered[['name', 'close', 'price_book_fq', 'enterprise_value_ebitda_ttm', 
             'current_ratio_fq', 'Recommend.All', 'analyst_score', 'analyst_rating', 'recommendation_total']]


📊 Cyclical 종목: 9개 중 0개 필터링됨
조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)


Empty DataFrame
Columns: [name, close, price_book_fq, enterprise_value_ebitda_ttm, current_ratio_fq, Recommend.All, analyst_score, analyst_rating, recommendation_total]
Index: []

## 전략 2: Growth (고성장형)

**목표:** 매출이 빠르게 늘면서, 성장성 대비 주가가 싼 기업

| 필터 | 조건 | 설명 |
|------|------|------|
| 매출 성장률 YoY | >= 20% | 전년 대비 고속 성장 |
| PEG 비율 | < 1 | 성장률 감안 시 저평가 |
| 부채비율 | < 150% | 금리 리스크 관리 |
| 애널리스트/기술 등급 | Buy 이상 | 전문가 평가 매수 이상 |


In [156]:
# Growth (고성장형) 스크리너
count_growth, df_growth = (
    Query()
    .select(
        'name', 'close', 'change', 'volume', 'market_cap_basic',
        'sector', 'industry',
        'total_revenue_yoy_growth_ttm',  # 매출 성장률 YoY
        'price_earnings_growth_ttm',     # PEG 비율
        'debt_to_equity_fq',             # 부채비율
        'recommendation_buy',
        'recommendation_over',
        'recommendation_hold',
        'recommendation_under',
        'recommendation_sell',
        'recommendation_total',
        'Recommend.All',                 # 기술 등급 (숫자)
        'earnings_per_share_diluted_yoy_growth_ttm', # EPS 성장률
    )
    .where(
        col('is_primary') == True,
        col('total_revenue_yoy_growth_ttm') >= 20,
        col('price_earnings_growth_ttm') < 1,
        col('price_earnings_growth_ttm') >= 0.1,
        col('debt_to_equity_fq') < 1.5,
        col('recommendation_total') >= MIN_ANALYST_COUNT,  # 최소 애널리스트 수
        col('Recommend.All') >= TECH_RATING_BUY,
    )
    .order_by('price_earnings_growth_ttm', ascending=True)
    .limit(1000)
    .get_scanner_data()
)

# 애널리스트 점수 계산
df_growth = calculate_analyst_score(df_growth)

# 애널리스트 평점 Buy 이상 필터링
df_growth = filter_by_analyst(df_growth, ANALYST_SCORE_BUY)

# 섹터 필터링
if not df_growth.empty and 'sector' in df_growth.columns:
    df_growth_filtered = df_growth[df_growth['sector'].isin(SECTORS['growth'])]
else:
    df_growth_filtered = df_growth

print(f"📊 Growth 종목: {count_growth}개 중 {len(df_growth_filtered)}개 필터링됨")
print(f"조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)")
df_growth_filtered[['name', 'close', 'total_revenue_yoy_growth_ttm', 'price_earnings_growth_ttm',
           'Recommend.All', 'analyst_score', 'analyst_rating', 'recommendation_total']]


📊 Growth 종목: 40개 중 9개 필터링됨
조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)


name   close  total_revenue_yoy_growth_ttm  price_earnings_growth_ttm  \
0   ADMA   19.72                     27.624850                   0.108595   
3   KRYS  217.46                     54.507739                   0.116387   
4   GDYN    9.49                     23.631829                   0.117798   
6   SEZL   64.44                     88.738124                   0.175563   
8   HRMY   37.00                     21.127649                   0.225383   
12  CPRX   23.10                     25.563214                   0.285656   
17  TMDX  140.10                     41.203720                   0.346565   
25   SEI   53.25                     92.331719                   0.524295   
37  INOD   59.65                     73.565611                   0.958706   

    Recommend.All  analyst_score analyst_rating  recommendation_total  
0        0.400000       2.000000     Strong Buy                     4  
3        0.603030       1.833333     Strong Buy                    12  
4        0.424242       2.000000     Strong Buy                     6  
6        0.112121       1.500000     Strong Buy                     4  
8        0.603030       1.545455     Strong Buy                    11  
12       0.512121       1.857143     Strong Buy                     7  
17       0.445455       1.461538     Strong Buy                    13  
25       0.603030       1.833333     Strong Buy                    12  
37       0.112121       1.800000     Strong Buy                     5

## 전략 3: Finance (금융/자산주)

**목표:** 극도로 저평가된 자산과 높은 자본효율, 배당 매력

| 필터 | 조건 | 설명 |
|------|------|------|
| PBR | < 0.6 | 절대적 저평가 영역 |
| ROE | >= 10% | 저평가지만 돈은 잘 버는 곳 |
| 배당수익률 | >= 4% | 확실한 현금 보상 |
| 애널리스트/기술 등급 | Buy 이상 | 전문가 평가 매수 이상 |


In [155]:
# Finance (금융/자산주) 스크리너
count_finance, df_finance = (
    Query()
    .select(
        'name', 'close', 'change', 'volume', 'market_cap_basic',
        'sector', 'industry',
        'price_book_fq',              # PBR
        'return_on_equity_fq',        # ROE
        'dividend_yield_recent',      # 배당수익률
        'recommendation_buy',
        'recommendation_over',
        'recommendation_hold',
        'recommendation_under',
        'recommendation_sell',
        'recommendation_total',
        'Recommend.All',              # 기술 등급 (숫자)
    )
    .where(
        col('is_primary') == True,
        col('price_book_fq') < 0.6,
        col('price_book_fq') > 0,
        col('return_on_equity_fq') >= 10,
        col('dividend_yield_recent') >= 4,
        col('recommendation_total') >= MIN_ANALYST_COUNT,  # 최소 애널리스트 수
        col('Recommend.All') >= TECH_RATING_BUY,
    )
    .order_by('dividend_yield_recent', ascending=False)
    .limit(1000)
    .get_scanner_data()
)

# 애널리스트 점수 계산
df_finance = calculate_analyst_score(df_finance)

# 애널리스트 평점 Buy 이상 필터링
df_finance = filter_by_analyst(df_finance, ANALYST_SCORE_BUY)

# 섹터 필터링 (Finance만)
if not df_finance.empty and 'sector' in df_finance.columns:
    df_finance_filtered = df_finance[df_finance['sector'].isin(SECTORS['finance'])]
else:
    df_finance_filtered = df_finance

print(f"📊 Finance 종목: {count_finance}개 중 {len(df_finance_filtered)}개 필터링됨")
print(f"조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)")
df_finance_filtered[['name', 'close', 'price_book_fq', 'return_on_equity_fq', 'dividend_yield_recent',
            'Recommend.All', 'analyst_score', 'analyst_rating', 'recommendation_total']]


📊 Finance 종목: 1개 중 0개 필터링됨
조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)


Empty DataFrame
Columns: [name, close, price_book_fq, return_on_equity_fq, dividend_yield_recent, Recommend.All, analyst_score, analyst_rating, recommendation_total]
Index: []

## 전략 4: Defensive (경기방어주)

**목표:** 마진이 안정적이고, 현금이 잘 돌며 배당을 주는 기업

| 필터 | 조건 | 설명 |
|------|------|------|
| 영업이익률 | >= 5% | 안정적인 마진 확보 |
| FCF | > 0 | 현금이 플러스인지 확인 |
| 배당수익률 | >= 3% | 은행 이자 이상의 수익 |
| 애널리스트/기술 등급 | Buy 이상 | 전문가 평가 매수 이상 |


In [157]:
# Defensive (경기방어주) 스크리너
count_defensive, df_defensive = (
    Query()
    .select(
        'name', 'close', 'change', 'volume', 'market_cap_basic',
        'sector', 'industry',
        'operating_margin_ttm',       # 영업이익률
        'free_cash_flow_ttm',         # 잉여현금흐름
        'dividend_yield_recent',      # 배당수익률
        'recommendation_buy',
        'recommendation_over',
        'recommendation_hold',
        'recommendation_under',
        'recommendation_sell',
        'recommendation_total',
        'Recommend.All',              # 기술 등급 (숫자)
    )
    .where(
        col('is_primary') == True,
        col('operating_margin_ttm') >= 5,
        col('free_cash_flow_ttm') > 0,
        col('dividend_yield_recent') >= 3,
        col('recommendation_total') >= MIN_ANALYST_COUNT,  # 최소 애널리스트 수
        col('Recommend.All') >= TECH_RATING_BUY,
    )
    .order_by('dividend_yield_recent', ascending=False)
    .limit(1000)
    .get_scanner_data()
)

# 애널리스트 점수 계산
df_defensive = calculate_analyst_score(df_defensive)

# 애널리스트 평점 Buy 이상 필터링
df_defensive = filter_by_analyst(df_defensive, ANALYST_SCORE_BUY)

# 섹터 필터링
if not df_defensive.empty and 'sector' in df_defensive.columns:
    df_defensive_filtered = df_defensive[df_defensive['sector'].isin(SECTORS['defensive'])]
else:
    df_defensive_filtered = df_defensive

print(f"📊 Defensive 종목: {count_defensive}개 중 {len(df_defensive_filtered)}개 필터링됨")
print(f"조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)")
df_defensive_filtered[['name', 'close', 'operating_margin_ttm', 'dividend_yield_recent',
              'Recommend.All', 'analyst_score', 'analyst_rating', 'recommendation_total']]


📊 Defensive 종목: 279개 중 6개 필터링됨
조건: 기술등급 Buy↑, 애널리스트 Buy↑ (score >= 0.5)


name  close  operating_margin_ttm  dividend_yield_recent  Recommend.All  \
100  TIGO  52.36             24.705041               5.704506       0.175758   
128   BIP  36.41             24.545863               4.775125       0.603030   
165   ELP  10.94             17.313574               4.081633       0.466667   
189  REYN  24.42             13.403452               3.789127       0.245455   
233  IRDM  17.13             26.745821               3.413174       0.157576   
252   KDP  28.33             21.979721               3.270530       0.133333   

     analyst_score analyst_rating  recommendation_total  
100       0.555556            Buy                     9  
128       1.500000     Strong Buy                    12  
165       2.000000     Strong Buy                    10  
189       0.750000            Buy                     8  
233       1.111111     Strong Buy                     9  
252       1.277778     Strong Buy                    18

## 스크리닝 결과 요약


In [160]:
# 결과 요약
print("=" * 60)
print("📊 4가지 투자 전략 스크리닝 결과 요약")
print("=" * 60)
print(f"  • Cyclical (경기민감형): {len(df_cyclical_filtered)}개 종목")
print(f"  • Growth (고성장형): {len(df_growth_filtered)}개 종목")
print(f"  • Finance (금융/자산주): {len(df_finance_filtered)}개 종목")
print(f"  • Defensive (경기방어주): {len(df_defensive_filtered)}개 종목")
print("=" * 60)

# 결과 저장 함수
def save_all_results():
    from datetime import datetime
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    
    results = {
        'cyclical': df_cyclical_filtered,
        'growth': df_growth_filtered,
        'finance': df_finance_filtered,
        'defensive': df_defensive_filtered
    }
    
    for strategy, df in results.items():
        if not df.empty:
            filename = f'global_{strategy}_{timestamp}.csv'
            df.to_csv(filename, index=False, encoding='utf-8-sig')
            print(f"✅ 저장: {filename}")

# 저장하려면 아래 주석 해제
save_all_results()


📊 4가지 투자 전략 스크리닝 결과 요약
  • Cyclical (경기민감형): 0개 종목
  • Growth (고성장형): 9개 종목
  • Finance (금융/자산주): 0개 종목
  • Defensive (경기방어주): 6개 종목
✅ 저장: global_growth_20251204_145534.csv
✅ 저장: global_defensive_20251204_145534.csv
